src/main/java/net/neoforged/neoforge/network/negotiation/NetworkComponentNegotiator.java

```diff
@@ -0,0 +1,182 @@
+        server.forEach(s -> finalClient.forEach(c -> {
+            if (s.id().equals(c.id())) {
+                matches.put(s.id(), s, c);
+            }
+        }));
+
+        client.removeIf(c -> matches.containsRow(c.id()));
+        server.removeIf(c -> matches.containsRow(c.id()));
+
+        if (!client.isEmpty()) {
+            final Map<ResourceLocation, Component> failureReasons = Maps.newHashMap();
+            client.forEach(c -> failureReasons.put(c.id(), Component.translatable("neoforge.network.negotiation.failure.missing.client.server")));
+            return new NegotiationResult(List.of(), false, failureReasons);
+        }
+
+        if (!server.isEmpty()) {
+            final Map<ResourceLocation, Component> failureReasons = Maps.newHashMap();
+            server.forEach(c -> failureReasons.put(c.id(), Component.translatable("neoforge.network.negotiation.failure.missing.server.client")));
+            return new NegotiationResult(List.of(), false, failureReasons);
+        }
+
+        final List<NegotiatedNetworkComponent> result = Lists.newArrayList();
+        final Map<ResourceLocation, Component> failureReasons = Maps.newHashMap();
+        for (Table.Cell<ResourceLocation, NegotiableNetworkComponent, NegotiableNetworkComponent> resourceLocationNegotiableNetworkComponentNegotiableNetworkComponentCell : matches.cellSet()) {
+            final NegotiableNetworkComponent serverComponent = resourceLocationNegotiableNetworkComponentNegotiableNetworkComponentCell.getColumnKey();
+            final NegotiableNetworkComponent clientComponent = resourceLocationNegotiableNetworkComponentNegotiableNetworkComponentCell.getValue();
+
+            Optional<ComponentNegotiationResult> serverToClientComparison = validateComponent(serverComponent, clientComponent, "client");
+            if (serverToClientComparison.isPresent() && !serverToClientComparison.get().success()) {
+                failureReasons.put(serverComponent.id(), serverToClientComparison.get().failureReason());
+                continue;
+            }
+
+            Optional<ComponentNegotiationResult> clientToServerComparison = validateComponent(clientComponent, serverComponent, "server");
+            if (clientToServerComparison.isPresent() && !clientToServerComparison.get().success()) {
+                failureReasons.put(serverComponent.id(), clientToServerComparison.get().failureReason());
+                continue;
+            }
+
+            //We can take the servers version here. Either both sides have the same version, or both sides have no version.
+            result.add(new NegotiatedNetworkComponent(serverComponent.id(), serverComponent.version()));
+        }
+
+        if (!failureReasons.isEmpty()) {
+            result.clear();
+        }
+        return new NegotiationResult(result, failureReasons.isEmpty(), failureReasons);
```

Comment:
Instead of explicitly clearing the result do we potentially want this to be something like:
```java
if (failureReasons.isEmpty()) {
    return new NegotiationResult(result, true, failureReasons);
}
return new NegotiationResult(List.of(), false, failureReasons);
```